src/aspire/abinitio/commonline_sync3n.py

```diff
@@ -0,0 +1,1025 @@
+        self.S_weighting = S_weighting
+        self.J_weighting = J_weighting
+        self._D_null = 1e-13
+        self.hist_intervals = int(hist_intervals)
+        # Warn if histogram may be too sparse for curve fitting
+        if self.S_weighting and (src.n < hist_intervals):
+            logger.warning(
+                f"`hist_intervals` {hist_intervals} > src.n {src.n}."
+                "  Consider reducing if curve fitting is infeasable."
+            )
+
+        # Auto configure GPU
+        self._gpu_module = None
+        try:
+            import cupy as cp
+
+            if cp.cuda.runtime.getDeviceCount() >= 1:
+                gpu_id = cp.cuda.runtime.getDevice()
+                logger.info(
+                    f"cupy and GPU {gpu_id} found by cuda runtime; enabling cupy."
+                )
+                self._gpu_module = self._init_cupy_module()
+            else:
+                logger.info("GPU not found, defaulting to numpy.")
+
+        except ModuleNotFoundError:
+            logger.info("cupy not found, defaulting numpy.")
+
+    ###########################################
+    # High level algorithm steps              #
+    ###########################################
+    def estimate_rotations(self):
+        """
+        Estimate rotation matrices.
+
+        :return: Array of rotation matrices, size n_imgx3x3.
+        """
+
+        # Initial estimate of viewing directions
+        Rijs0 = self._estimate_relative_viewing_directions()
+
+        # Compute and apply global handedness
+        Rijs = self._global_J_sync(Rijs0)
+
+        # Build sync3n matrix
+        S = self._construct_sync3n_matrix(Rijs)
+
+        # Optionally compute S weights
+        W = None
+        if self.S_weighting is True:
+            W = self._syncmatrix_weights(Rijs)
+
+        # Yield rotations from S
+        self.rotations = self._sync3n_S_to_rot(S, W)
+
+    #######################
+    # Main Sync3N Methods #
+    #######################
+    def _sync3n_S_to_rot(self, S, W=None, n_eigs=4):
+        """
+        Use eigen decomposition of S to estimate transforms,
+        then project transforms to nearest rotations.
+
+        :param S: Numpy array represeting Synchronization matrix.
+        :param W: Optional weights array, default `None` is equal weighting of `S`.
+        :param n_eigs: Optional, number of eigenvalues to compute (min 3).
+        """
+
+        if n_eigs < 3:
+            raise ValueError(
+                f"n_eigs must be greater than 3, default is 4. Invoked with {n_eigs}"
+            )
+
+        if W is not None:
+            logger.info("Applying weights to synchronization matrix.")
+            if not W.shape == (self.n_img, self.n_img):
+                raise RuntimeError(
+                    f"Shape of W should be {(self.n_img, self.n_img)}."
+                    f" Received {W.shape}."
+                )
+            # Initialize D
+            D = np.mean(W, axis=1)  # D, check axis
```

Comment:
Yeah, was a note to myself. Should be. I'll remove.

```
% Prepare weights for the blocks of the matrix S.
D = mean(W,2); % We use mean and not sum so that after normalizing W the 
               % sum of its rows will be N, as required.
```